src/vmm/src/guest_config/templates/mod.rs

```diff
@@ -0,0 +1,1043 @@
+            match character {
+                '1' => bitmask_str.push(value_str.as_bytes()[idx] as char),
+                _ => bitmask_str.push('x'),
+            }
+        }
+
+        debug!(
+            "Serializing:\n{}\n",
+            format!(
+                "Input:\nFilter: \t[{}]\nValue: \t\t[{}]\nOutput:\nBitmask: \t[{}]\n",
+                filter_str,
+                value_str,
+                bitmask_str.as_str().replace("0b", ""),
+            )
+        );
+
+        serializer.serialize_str(bitmask_str.as_str())
+    }
+}
+
+/// Guest config sub-module specifically for
+/// config templates.
+#[cfg(target_arch = "aarch64")]
+pub mod aarch64 {
+    use std::str::FromStr;
+
+    use log::debug;
+    use serde::de::Error as SerdeError;
+    use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+    use crate::guest_config::aarch64::Aarch64CpuConfiguration;
+    use crate::guest_config::templates::{deserialize_u64_from_str, Error};
+
+    /// Wrapper type to containing aarch64 CPU config modifiers.
+    #[derive(Debug, Deserialize, Eq, PartialEq, Serialize)]
+    pub struct Aarch64CpuTemplate {
+        /// Modifiers for registers on Aarch64 CPUs.
+        pub reg_modifiers: Vec<RegisterModifier>,
+    }
+
+    /// Bit-mask based wrapper to apply
+    /// changes to a given register's value
+    #[derive(Debug, Deserialize, Eq, PartialEq, Serialize)]
+    pub struct RegisterModifier {
+        /// Pointer of the location to be bit masked.
+        #[serde(
+            deserialize_with = "deserialize_u64_from_str",
+            serialize_with = "serialize_u64_to_hex_str"
+        )]
+        pub addr: u64,
+        /// Mask to be applied to register's value at the address
+        /// provided.
+        #[serde(
+            deserialize_with = "deserialize_u128_bitmask",
+            serialize_with = "serialize_u128_bitmask"
+        )]
+        pub mask: BitMask,
+    }
+
+    /// Bit mask to adjust targeted bits to the value,
+    /// applied by the filter only.
+    #[derive(Debug, Deserialize, Eq, PartialEq, Serialize)]
+    pub struct BitMask {
+        /// Mask value is to be applied according
+        /// to what is allowed through by the filter.
+        #[serde(deserialize_with = "deserialize_u128_from_str")]
+        pub filter: u128,
+        /// Mask value to be applied.
+        #[serde(deserialize_with = "deserialize_u128_from_str")]
+        pub value: u128,
+    }
+
+    /// CPU template with modifiers to be applied on
+    /// top of an existing configuration to generate
+    /// the guest configuration to be used.
+    pub fn create_guest_cpu_config(
+        _template: &Aarch64CpuTemplate,
+        _host_config: &Aarch64CpuConfiguration,
+    ) -> Result<Aarch64CpuConfiguration, Error> {
+        // TODO
+        Ok(Aarch64CpuConfiguration::default())
+    }
+
+    fn deserialize_u128_from_str<'de, D>(deserializer: D) -> Result<u128, D::Error>
+    where
```

Comment:
The numerical JSON ops are quite duplicated. They are split out in this way for the moment because each arch has its own bit-widths essentially. I wanted to introduce more generic types to galvanize the code, but it made the types quite hard to process for a first pass. A TODO for now seems reasonable.

Reply:
How will we be tracking this? A TODO in the code or something else?